Allow longer name for AWS/Azure cluster and refactoring

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

This PR
1. fixes the first part of the TODOs in #963, to unblock the user from using a long name for spot jobs on AWS. 
2. moves the cluster name checking into each Cloud.
3. refactors the `cloud.supports` to `cloud.check_features_are_supported` and change the boolean return value to exception raising.
4. optimizes the failover for `InvalidClusterNameError` and `ClusterUserIdentityError` to directly move to the next cloud, as those errors apply to the whole cloud.
5. fixes minor UX issues for failover:
  Previously:
    <img width="829" alt="image" src="https://user-images.githubusercontent.com/6753189/215648917-c6fbafba-337a-4047-a56b-0992fcef3b63.png">
  Now:
    <img width="824" alt="image" src="https://user-images.githubusercontent.com/6753189/215648995-7c3bf83e-5833-4145-8d0b-566a8c137972.png">



<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-135-chars --cloud aws`
    - [x] AWS 244 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-by-checking-the-document-of-aws-we-found-that-the-cluster-name-length-should-be-less-than-250-due-to-the-tag-limit-this-sentence-has-246-c --num-nodes 2 --cloud aws -t t3.micro` (and 247 does not work)
    - [x] GCP (didn't change the original tested limit)
    - [x] Azure 42 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluste --cloud azure --region southcentralus --num-nodes 2`
    - [x]  Lambda 64 characters: `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-mor --cloud lambda` works
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-azure-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-136-chars --cloud azure`: correctly fails for the Azure
    - [x] `sky launch -c i-am-using-a-very-long-name-for-the-aws-cluster-it-should-be-more-than-63-characters-and-hopefully-it-works-this-sentence-has-135-chars --gpus A100:8`, it successfully and quickly failover through Lambda/GCP/Azure and AWS (skip the first three clouds with a single try and try all the regions on AWS).

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -66,6 +68,14 @@ class Cloud:
 
     _REPR = '<Cloud>'
 
+    _MAX_CLUSTER_NAME_LEN_LIMIT: Optional[int] = None
+
+    # All features implemented by this cloud. Used by
+    # Cloud.check_features_are_supported()
+    # None means all features are supported.
+    _CLOUD_IMPLEMENTATION_FEATURES: Optional[
+        Set[CloudImplementationFeatures]] = None
```
For discussion: I feel putting these as methods will more clearly remind implementors that they need to be filled in. Here, the naming of the vars suggests they are "constants", which may be neglected?

concretevitamin replied:
Should None default to mean none supported? (Or, we can change default val to set().) 

Because we may add new values (for new features) to CloudImplementationFeatures in the future and that doesn't mean existing Clouds that default to None automatically support those.

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
-                 requested_features: Set[CloudImplementationFeatures]) -> bool:
-        """Returns whether all of the requested features are supported.
+    def check_features_are_supported(
+            cls, requested_features: Set[CloudImplementationFeatures]):
```
nit: Shall we consistently -> None?

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
 
         For instance, Lambda Cloud does not support autostop, so
-        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) returns False.
+        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) raises the
```
need update

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
+                [f.name for f in unsupported_features])
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.NotSupportedError(
+                    f'Cloud {cls._REPR} does not support the following '
```
See another comment?

Michaelvll replied:
Please check https://github.com/skypilot-org/skypilot/pull/1648#discussion_r1091529257

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
+                    f'features: {unsupported_features_str}')
+
+    @classmethod
+    def check_cluster_name_is_valid(cls, cluster_name: str):
```
ditto

Michaelvll replied:
Sorry, what does this mean?

concretevitamin replied:
Can we add `-> None`?

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
+
+        Bans (including but not limited to) names that:
+        - are digits-only
+        - contain underscore (_)
```
```suggestion
        - contain underscore (_)
        
```

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +330,55 @@ def need_cleanup_after_preemption(self,
+            with ux_utils.print_exception_no_traceback():
+                raise exceptions.InvalidClusterNameError(
+                    f'Cluster name "{cluster_name}" is invalid; '
+                    f'ensure it is fully matched by regex: {valid_regex}')
```
```suggestion
                    f'ensure it is fully matched by regex (e.g., no underscores or capital letters): {valid_regex}')
```
minor, not feeling strongly: seen such cases

Michaelvll replied:
Added. Thanks!

concretevitamin commented on sky/cli.py:
```diff
@@ -3077,7 +3077,7 @@ def spot_launch(
         name = backend_utils.generate_cluster_name()
     else:
-        backend_utils.check_cluster_name_is_valid(name)
+        clouds.Cloud.check_cluster_name_is_valid(name)
```
Maybe comment that this does a basic regex check, while name length check will be done when the controller has decided which cloud to use and begins provisioning the spot cluster?

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -1699,6 +1693,17 @@ def provision_with_retries(
+                # InvalidClusterNameError: cluster name is invalid,
+                # NotSupportedError: cloud does not support requested features,
+                # CloudUserIdentityError: cloud user identity is invalid.
+                # The exceptions above should not be applicable to the whole
```
```suggestion
                # The exceptions above should be applicable to the whole
```

concretevitamin commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -3198,11 +3197,12 @@ def _check_existing_cluster(
-        backend_utils.check_cluster_name_is_valid(cluster_name, task_cloud)
+        resources = list(task.resources)[0]
+        task_cloud = (resources.cloud
+                      if resources.cloud is not None else clouds.Cloud)
```
Would resources.cloud ever be None here?

concretevitamin reviewed (commented):
Thanks @Michaelvll, and nice tests! Shall we also test a name length close to AWS's char limit?

Michaelvll commented on sky/core.py:
```diff
@@ -312,9 +312,8 @@ def stop(cluster_name: str, purge: bool = False) -> None:
     cloud = handle.launched_resources.cloud
-    if not cloud.supports({clouds.CloudImplementationFeatures.STOP}):
-        raise exceptions.NotSupportedError(
-            (f'{cloud} does not support stopping instances.'))
```
Good point! I changed the implementation by having each cloud setting the unsupported features along with the reason. Wdyt?

concretevitamin replied:
Can we add a note to CloudImplementationFeatures that if any value (feature) is added to that enum, audit all Cloud implementations and potentially add that feature to _cloud_unsupported_features()?

This is related to another comment. However, I do see that to provide a reason message for each unsupported feature this current approach seems appropriate.

Michaelvll replied:
Good point! Added. Thanks!

Michaelvll commented:
> Shall we also test a name length close to AWS's char limit?

Good catch! Actually, we have to reduce the limit to 246 for AWS because of the DescribeInstance API.

concretevitamin commented on sky/clouds/aws.py:
```diff
@@ -65,6 +66,15 @@ class AWS(clouds.Cloud):
+        return dict()
+
+    @classmethod
+    def _max_cluster_name_len_limit(cls) -> Optional[int]:
```
```suggestion
    def _max_cluster_name_length(cls) -> Optional[int]:
```

concretevitamin commented on sky/clouds/lambda_cloud.py:
```diff
@@ -16,18 +16,36 @@
     """Lambda Labs GPU Cloud."""
 
     _REPR = 'Lambda'
+    # Lamdba has a 64 char limit for cluster name.
```
nit: a blank line between definitions?

concretevitamin commented on sky/clouds/lambda_cloud.py:
```diff
@@ -16,18 +16,36 @@
     _REPR = 'Lambda'
+    # Lamdba has a 64 char limit for cluster name.
+    # Reference: https://cloud.lambdalabs.com/api/v1/docs#operation/launchInstance # pylint: disable=line-too-long
+    _MAX_CLUSTER_NAME_LEN_LIMIT = 64
```
For all such Cloud's own limits, don't the `ray-` prefix and suffix occupy some chars? Can we test close to the limit for all 4 clouds?

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -320,14 +348,62 @@ def need_cleanup_after_preemption(self,
 
         For instance, Lambda Cloud does not support autostop, so
-        Lambda.support({CloudImplementationFeatures.AUTOSTOP}) returns False.
+        Lambda.check_features_are_supported({
```
func name needs update

concretevitamin commented on sky/clouds/lambda_cloud.py:
```diff
@@ -16,18 +16,36 @@
+    _CLOUD_UNSUPPORTED_FEATURES = {
+        clouds.CloudImplementationFeatures.STOP: 'Lambda cloud does not support stopping VMs.',
+        clouds.CloudImplementationFeatures.AUTOSTOP: 'Lambda cloud does not support stopping VMs.',
+        clouds.CloudImplementationFeatures.MULTI_NODE: 'Multi-node is not supported by the implementation yet.',
```
```suggestion
        clouds.CloudImplementationFeatures.MULTI_NODE: 'Multi-node is not supported by the Lambda Cloud implementation yet.',
```

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -66,6 +69,31 @@ class Cloud:
+    def _max_cluster_name_len_limit(cls) -> Optional[int]:
+        """Returns the maximum length limit of a cluster name.
+
+        This method is used by check_cluster_name() to check if the cluster
```
```suggestion
        This method is used by check_cluster_name_is_valid() to check if the cluster
```

concretevitamin reviewed (approved):
Thanks @Michaelvll. Just some minor comments and a question about some additional characters added by the autoscaler. May be worth updating PR description with each cloud's at-limit test?

Michaelvll commented:
> May be worth updating PR description with each cloud's at-limit test?

Great point! I added the tests for each cloud, and seems there are multiple corner cases that are caught by those tests. I fixed the cluster name limit for AWS and Azure. Thanks!
